File: compat/suites/python-sdk/groups/kinesis.py
"""
groups/kinesis.py — Kinesis compatibility test implementations.
"""

from __future__ import annotations
import base64
import time
from lib.harness import TestContext
from lib.clients import make_clients


def _kin(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).kinesis


def _wait_stream_active(kin, stream_name: str, max_wait: int = 10) -> None:
    deadline = time.time() + max_wait
    while time.time() < deadline:
        resp = kin.describe_stream_summary(StreamName=stream_name)
        status = resp["StreamDescriptionSummary"]["StreamStatus"]
        if status == "ACTIVE":
            return
        time.sleep(0.2)
    raise TimeoutError(f"Stream {stream_name!r} did not become ACTIVE within {max_wait}s")


# ── kinesis-records ───────────────────────────────────────────────────────────

def setup_kinesis_records(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = f"oc-{ctx.run_id}-rec"
    kin.create_stream(StreamName=name, ShardCount=1)
    _wait_stream_active(kin, name)
    ctx["kinesis_rec_stream"] = name


def teardown_kinesis_records(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx.get("kinesis_rec_stream")
    if not name:
        return
    try:
        kin.delete_stream(StreamName=name, EnforceConsumerDeletion=True)
    except Exception:
        pass


def PutRecord(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_rec_stream"]
    resp = kin.put_record(
        StreamName=name,
        Data=b"hello-record",
        PartitionKey="pk-1",
    )
    if not resp.get("ShardId"):
        raise AssertionError(f"PutRecord: missing ShardId in {resp}")
    ctx["kinesis_shard_id"] = resp["ShardId"]


def PutRecords(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_rec_stream"]
    records = [
        {"Data": f"record-{i}".encode(), "PartitionKey": f"pk-{i}"}
        for i in range(5)
    ]
    resp = kin.put_records(StreamName=name, Records=records)
    if resp.get("FailedRecordCount", 0) > 0:
        raise AssertionError(f"PutRecords: {resp['FailedRecordCount']} failed records")


def GetShardIterator(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_rec_stream"]
    shard_id = ctx.get("kinesis_shard_id") or "shardId-000000000000"
    resp = kin.get_shard_iterator(
        StreamName=name,
        ShardId=shard_id,
        ShardIteratorType="TRIM_HORIZON",
    )
    it = resp.get("ShardIterator")
    if not it:
        raise AssertionError("GetShardIterator: missing ShardIterator")
    ctx["kinesis_shard_iter"] = it


def GetRecords(ctx: TestContext) -> None:
    kin = _kin(ctx)
    it = ctx.get("kinesis_shard_iter")
    if not it:
        raise AssertionError("GetRecords: no shard iterator in context (run GetShardIterator first)")
    resp = kin.get_records(ShardIterator=it, Limit=100)
    records = resp.get("Records", [])
    if not records:
        raise AssertionError("GetRecords: no records returned (expected ≥1 from PutRecord/PutRecords)")
    # Verify data is readable
    _ = base64.b64decode(records[0]["Data"]) if isinstance(records[0]["Data"], str) else records[0]["Data"]


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "kinesis-records:PutRecord": PutRecord,
    "kinesis-records:PutRecords": PutRecords,
    "kinesis-records:GetShardIterator": GetShardIterator,
    "kinesis-records:GetRecords": GetRecords,
}

SETUP = {
    "kinesis-records": setup_kinesis_records,
}

TEARDOWN = {
    "kinesis-records": teardown_kinesis_records,
}

File: compat/suites/go-sdk/internal/groups/kinesis.go
package groups

import (
	"context"
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/kinesis"
	"github.com/aws/aws-sdk-go-v2/service/kinesis/types"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/clients"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/harness"
)

func Kinesis(c *clients.Clients) ServiceGroup {
	g := &kinesisGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"kinesis-records:PutRecord":        g.PutRecord,
			"kinesis-records:PutRecords":       g.PutRecords,
			"kinesis-records:GetShardIterator": g.GetShardIterator,
			"kinesis-records:GetRecords":       g.GetRecords,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"kinesis-records": g.setupRecords,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"kinesis-records": g.teardownRecords,
		},
	}
}

type kinesisGroup struct{ c *clients.Clients }

func (g *kinesisGroup) cl() *kinesis.Client { return g.c.Kinesis() }

func (g *kinesisGroup) waitStreamActive(ctx context.Context, streamName string) error {
	deadline := time.Now().Add(30 * time.Second)
	for time.Now().Before(deadline) {
		resp, err := g.cl().DescribeStreamSummary(ctx, &kinesis.DescribeStreamSummaryInput{
			StreamName: aws.String(streamName),
		})
		if err != nil {
			return err
		}
		if resp.StreamDescriptionSummary.StreamStatus == types.StreamStatusActive {
			return nil
		}
		time.Sleep(500 * time.Millisecond)
	}
	return fmt.Errorf("stream %q did not become ACTIVE within timeout", streamName)
}

// ── kinesis-records ───────────────────────────────────────────────────────────

func (g *kinesisGroup) setupRecords(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("oc-rec-%s", t.RunID)
	if _, err := g.cl().CreateStream(ctx, &kinesis.CreateStreamInput{
		StreamName: aws.String(name), ShardCount: aws.Int32(1),
	}); err != nil {
		return err
	}
	if err := g.waitStreamActive(ctx, name); err != nil {
		return err
	}
	t.Set("kinesis_rec_stream", name)
	return nil
}

func (g *kinesisGroup) teardownRecords(ctx context.Context, t *harness.TestContext) error {
	if name := t.GetString("kinesis_rec_stream"); name != "" {
		g.cl().DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(name)}) //nolint:errcheck
	}
	return nil
}

func (g *kinesisGroup) PutRecord(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_rec_stream")
	resp, err := g.cl().PutRecord(ctx, &kinesis.PutRecordInput{
		StreamName:   aws.String(stream),
		Data:         []byte("record-data"),
		PartitionKey: aws.String("pk1"),
	})
	if err != nil {
		return err
	}
	if aws.ToString(resp.ShardId) == "" {
		return fmt.Errorf("PutRecord: missing ShardId in response")
	}
	return nil
}

func (g *kinesisGroup) PutRecords(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_rec_stream")
	resp, err := g.cl().PutRecords(ctx, &kinesis.PutRecordsInput{
		StreamName: aws.String(stream),
		Records: []types.PutRecordsRequestEntry{
			{Data: []byte("r1"), PartitionKey: aws.String("pk1")},
			{Data: []byte("r2"), PartitionKey: aws.String("pk2")},
		},
	})
	if err != nil {
		return err
	}
	if aws.ToInt32(resp.FailedRecordCount) > 0 {
		return fmt.Errorf("PutRecords: %d records failed", aws.ToInt32(resp.FailedRecordCount))
	}
	return nil
}

func (g *kinesisGroup) GetShardIterator(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_rec_stream")
	// Get first shard ID
	desc, err := g.cl().DescribeStream(ctx, &kinesis.DescribeStreamInput{StreamName: aws.String(stream)})
	if err != nil {
		return err
	}
	if len(desc.StreamDescription.Shards) == 0 {
		return fmt.Errorf("GetShardIterator: no shards")
	}
	shardID := aws.ToString(desc.StreamDescription.Shards[0].ShardId)
	resp, err := g.cl().GetShardIterator(ctx, &kinesis.GetShardIteratorInput{
		StreamName:        aws.String(stream),
		ShardId:           aws.String(shardID),
		ShardIteratorType: types.ShardIteratorTypeTrimHorizon,
	})
	if err != nil {
		return err
	}
	t.Set("kinesis_shard_iter", aws.ToString(resp.ShardIterator))
	return nil
}

func (g *kinesisGroup) GetRecords(ctx context.Context, t *harness.TestContext) error {
	iter := t.GetString("kinesis_shard_iter")
	if iter == "" {
		return nil
	}
	resp, err := g.cl().GetRecords(ctx, &kinesis.GetRecordsInput{
		ShardIterator: aws.String(iter),
		Limit:         aws.Int32(10),
	})
	if err != nil {
		return err
	}
	if len(resp.Records) == 0 {
		return fmt.Errorf("GetRecords: expected ≥1 record")
	}
	return nil
}
